Allows the release namespace to be overridden

KhizerJaan opened the pull request:
**Problem:**
- Vault can only be deployed in Release namespace.

**Root Cause:**
- namespace is hardcoded to `{{ .Release.Namespace }}`

**Solution:**
- Templatize namespaec in _helpers.tpl

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/vault-helm?pullRequest=909)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/vault-helm?pullRequest=909)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/vault-helm?pullRequest=909).</sub>

tomhjp commented:
Hi @KhizerJaan - please could you provide some more context about why this is useful?

KhizerJaan commented:
If Vault Helm Chart is being used as dependency in another helm (umbrella) chart, it is deployed to umbrella chart release namespace . This PR would enabled to override namespace for Vault Helm Chart.

KhizerJaan commented:
@tomhjp kindly review and merge the pull request at your convenience. I have umbrella chart use case and am blocked by it . 
By allowing users to choose the deployment namespace, Helm charts become more versatile and adaptable to different environments and deployment strategies. Users can deploy the chart in the namespace that aligns with their specific requirements or organizational practices .

tomhjp commented:
Thanks for the explanation, that makes sense. This seems like a nice improvement to merge then with a few additions. Please could we add some bats unit tests for each resource, a default empty value in the Values file, an entry in the schema, and a short changelog entry under unreleased improvements? Happy to help out with any of those, just let me know if you need some pointers or assistance.

KhizerJaan commented:
Sure. Highly appreciate the feedback and suggestions @tomhjp . I will incorporate the suggested changes, and submit a revised version that addresses the feedback provided.

KhizerJaan commented:
@tomhjp I have incorporated the suggestions provided.
Kindly review the updated PR at your convenience. I believe that it is now complete and addresses all the necessary improvements. However, I welcome any further feedback or suggestions you may have to enhance the quality of the contribution.

KhizerJaan commented:
@tomhjp  Could you please consider merging this PR? Alternatively, I would greatly appreciate any feedback you might have.

tomhjp commented on values.yaml:
```diff
@@ -3,6 +3,10 @@
 
 # Available parameters and their default values for the Vault chart.
 
+# Allows to specify namespace other than Release.Namespace
+# If empty, default namespace would be Release.Namespace
```
```suggestion
# The namespace to deploy to. Defaults to the `helm` installation namespace.
```

tomhjp commented on values.yaml:
```diff
@@ -3,6 +3,10 @@
 
+# Allows to specify namespace other than Release.Namespace
+# If empty, default namespace would be Release.Namespace
+namespaceOverride: ""
```
I think this probably fits best in the `global` section just below.

tomhjp replied:
WDYT about just `namespace`? It feels descriptive enough to me.

```suggestion
namespace: ""
```

tomhjp commented on CHANGELOG.md:
```diff
@@ -44,9 +44,11 @@ Features:
 
 Improvements:
 * injector: Add `ephemeralLimit` and `ephemeralRequest` as options for configuring Agent's ephemeral storage resources [GH-798](https://github.com/hashicorp/vault-helm/pull/798)
+* Add `namespaceOverride` to specify namespace from values or command line. [GH-909](https://github.com/hashicorp/vault-helm/pull/909)
```
```suggestion
* global: Add `global.namespace` to override the helm installation namespace. [GH-909](https://github.com/hashicorp/vault-helm/pull/909)
```

tomhjp commented on test/unit/csi-agent-configmap.bats:
```diff
@@ -21,6 +21,25 @@ load _helpers
   [ "${actual}" = "release-name-vault-csi-provider-agent-config" ]
 }
 
+@test "csi/Agent-ConfigMap: namespace" {
```
Thanks for adding these tests, they look great.

tomhjp reviewed (commented):
Thanks for the changes so far - just a few more small suggestions and then I think this is ready to merge.

tomhjp commented:
I ran the CI and a few unit test failures came up FYI:

```text
not ok 166 injector/certs-secret: namespace is set in 131ms
not ok 168 injector/role: namespace is set in 130ms
not ok 170 injector/rolebinding: namespace is set in 135ms
not ok 174 injector/MutatingWebhookConfiguration: namespace is set in 138ms
```

KhizerJaan commented:
@tomhjp Thank you for suggestions . highly appreciate it . I've updated regards to your suggestions . PTAL

KhizerJaan commented:
@tomhjp PTAL

KhizerJaan commented:
@tomhjp can this be merged please.

tomhjp commented on values.yaml:
```diff
@@ -3,11 +3,15 @@
 
 # Available parameters and their default values for the Vault chart.
 
+
```
```suggestion
```

tomhjp reviewed (approved):
Sorry, the changelog changed in main since my last review - please could you rebase/merge in main to make sure the changes stay in the unreleased section when this is merged? Otherwise LGTM

KhizerJaan commented:
@tomhjp I can't merge this . can you please merge.

KhizerJaan commented:
@tomhjp can this be merged please .

KhizerJaan commented:
@tomhjp  branch synced . PTAL

KhizerJaan commented:
@tomhjp can you please merge this PR? I've synced with main . or it would be great if yo have any feedback.

tomhjp commented:
I've just run the tests again to prepare for merging, but it looks like the same ones from before are still failing. Do you need some help getting going with running them locally?

```
not ok 166 injector/certs-secret: namespace is set
not ok 168 injector/role: namespace is set
not ok 170 injector/rolebinding: namespace is set
not ok 174 injector/MutatingWebhookConfiguration: namespace is set
```

KhizerJaan commented:
@tomhjp  previously results were escaped like this `\"bar\"` and it was failing locally so i removed escaping characters . now I've reverted back to escaping . can you please test now?

tomhjp commented:
Looks like that's done the trick, thanks. The other way to fix it would be to add `-r` to the `yq` command, like `yq -r '.webhooks[0].clientConfig.service.namespace'`, but what you've got now LGTM.
